### PR TITLE
[WIP] Video labels

### PR DESCRIPTION
**Pull request by benjaminpkane**
## TLDR

Just using a reference field to a collection of video labels will not solve mongodb's document size limitations (16MB). I am proposing we model video labels as `FrameLabels` so document count (not size) grows as the number of labeled "frames" in a video dataset grows.

Or we take a short term solution of using mongodb's gridfs (or the filesystem) to store ETA's video labels and make them look like fields. Somewhat tricky, but less code. Will likely not "stick".

I could go either way. It is not the technical details that are important right now, but I think we should be aware of the limitations and technical debt we are signing up for. More below.

## Progress

Adds `fof.VideoLabels` class that directly inherits from ETA `VideoLabels`, and the concept of an `mtype` (media type) for `Sample`, `SampleView`, `Dataset`. `mtype` was inspired by numpy's `dtype`.

### mtype

`mtype` is essentially an enum; `image` or `video`.

Rules:
 * A sample always has a defined `mtype` because it is imputed from the required `filepath`
 * A sample's `mtype` cannot change, though its `filepath` can
 * A dataset's `mtype` is assigned as soon as it can be (when the first sample is added)
 * A dataset's `mtype` cannot be changed once assigned. Perhaps this doesn't need to be the case, but the alternative would be harder to support.

Most of this is handled currently, but there is likely more work to be done.

### Video labels

The important part of this work (video labels support) is still quite bare. What is done currently uses `gridfs` which is basically just a file store. It builds of off #424, but that may not be what we want. I am confident, though, that this can functional before the next release. It may be hard to move forward with it afterwards.

```py
import fiftyone as fo

labels = fo.VideoLabels.from_json("/path/to/labels.json")
sample = fo.Sample(filepath="/path/to/video.mp4")
dataset = fo.Dataset()
sample["labels"] = labels
dataset.add_sample(sample)
```

An alternative would be to close #424 and return to optimizing when more expertise and resources are available to us. We could then use MongoEngine to model video labels.

e.g.
```py
class FramesField(mongoengine.fields.MapField): # a field type only available to video samples

    def __init__(self, *args, **kwargs):
        return super().__init__(
            field=mongoengine.fields.ReferenceField("FrameSample"))

class FrameSample(DynamicDocument): # implies a separate collection for each `FramesField` created for a dataset

    meta = {"abstract": True}
``` 

There are some details missing here, but all labels/fields currently available to image samples would be assignable fields in a `FrameSample` document.

Happy to take this offline.

### CODE REVIEW

**Comment by benjaminpkane** [comment]
Closing. Work is now a part of #559.
